engine/crates/async-runtime/src/stream.rs

```diff
@@ -0,0 +1,49 @@
+use futures_util::{
+    future::{self, BoxFuture},
+    stream::{self, BoxStream},
+    Future, FutureExt, Stream, StreamExt,
+};
+
+/// Combines a stream and a "producer" future into a single stream
+///
+/// This can be used when you have the receivng side of a channel and a future that sends
+/// on that channel - combining the two into a single stream that'll run till the channel
+/// is exhausted.  If you drop the stream you also cancel the underlying process.
+pub fn producer_stream<'a, S, P, Item>(stream: S, producer: P) -> impl Stream<Item = Item> + 'a
```

Comment:
Any name suggestions for this welcome, it's an awkward concept to describe succinctly

Reply:
I think I would create an extension like:
```
impl Stream {
  fn with_producer_future(&self, producer: P)
}
```

Reply:
Or maybe `join_with_future`

Reply:
"join" feels a bit more idiomatic imho